gitlab/gitlab_hook.py

```diff
@@ -0,0 +1,678 @@
+
+    name: str
+    username: str
+    avatar_url: str
+
+
+@dataclass
+class GitlabProject(SerializableAttrs['GitlabProject']):
+
+    id: Optional[int] = None
+    name: Optional[str] = None
+    description: Optional[str] = None
+    web_url: Optional[str] = None
+    avatar_url: Optional[str] = None
+    git_ssh_url: Optional[str] = None
+    git_http_url: Optional[str] = None
+    namespace: Optional[str] = None
+    visibility_level: Optional[int] = None
+    path_with_namespace: Optional[str] = None
+    default_branch: Optional[str] = None
+    homepage: Optional[str] = None
+    url: Optional[str] = None
+    ssh_url: Optional[str] = None
+    http_url: Optional[str] = None
+
+
+@dataclass
+class GitlabCommit(SerializableAttrs['GitlabCommit']):
+
+    id: str
+    message: str
+    timestamp: Optional[datetime] = None
+    url: Optional[str] = None
+    author: Optional[str] = None
+    added: Optional[List[str]] = None
+    modified: Optional[List[str]] = None
+    removed: Optional[List[str]] = None
+
+
+@dataclass
+class GitlabAuthor(SerializableAttrs['GitlabAuthor']):
+
+    name: str
+    email: str
+
+
+@dataclass
+class GitlabRepository(SerializableAttrs['GitlabRepository']):
+
+    name: str
+    url: Optional[str] = None
+    description: Optional[str] = None
+    homepage: Optional[str] = None
+    git_http_url: Optional[str] = None
+    git_ssh_url: Optional[str] = None
+    visibility_level: Optional[int] = None
+
+
+@dataclass
+class GitlabStDiff(SerializableAttrs['GitlabStDiff']):
+
+    diff: str
+    new_path: str
+    old_path: str
+    a_mode: str
+    b_mode: str
+    new_file: bool
+    renamed_file: bool
+    deleted_file: bool
+
+
+@dataclass
+class GitlabSource(SerializableAttrs['GitlabSource']):
+
+    name: str
+    namespace: str
+
+    description: Optional[str] = None
+    web_url: Optional[str] = None
+    avatar_url: Optional[str] = None
+    git_ssh_url: Optional[str] = None
+    git_http_url: Optional[str] = None
+    visibility_level: Optional[int] = None
+    path_with_namespace: Optional[str] = None
+    default_branch: Optional[str] = None
+    homepage: Optional[str] = None
+    url: Optional[str] = None
+    ssh_url: Optional[str] = None
+    http_url: Optional[str] = None
+
+
+@dataclass
+class GitlabTarget(SerializableAttrs['GitlabTarget']):
+
+    name: str
+    namespace: str
+
+    description: Optional[str] = None
+    web_url: Optional[str] = None
+    avatar_url: Optional[str] = None
+    git_ssh_url: Optional[str] = None
+    git_http_url: Optional[str] = None
+    visibility_level: Optional[int] = None
+    path_with_namespace: Optional[str] = None
+    default_branch: Optional[str] = None
+    homepage: Optional[str] = None
+    url: Optional[str] = None
+    ssh_url: Optional[str] = None
+    http_url: Optional[str] = None
+
+
+@dataclass
+class GitlabChange(SerializableAttrs['GitlabChange']):
+
+    previous: Union[GitlabLabel, str, int]
+    current: Union[GitlabLabel, str, int]
```

Comment:
I think the deserializer currently doesn't support arbitrary unions, only `Union[type, None]` (aka `Optional[type]`). I could probably make it support arbitrary unions though if gitlab really sends different types.

Reply:
yes gitlab sends different types and is inconsistent everywhere. at the moment it doesn't brake anything as it seems to be parsed as a `mautrix.client.api.types.util.obj.Obj` (for a label change that is) but it would probably be nice to have the possibility of parsing Unions